### PR TITLE
Autonity contract integration

### PR DESCRIPTION
**Pull request by b00ris**
Closes https://github.com/clearmatics/autonity/issues/191

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -198,16 +185,14 @@ func (c *CliqueConfig) String() string {
...
+//// IstanbulConfig is the consensus engine configs for Istanbul based sealing.
 type IstanbulConfig struct {
-	Epoch          uint64         `json:"epoch"`             // Epoch length to reset votes and checkpoint
-	ProposerPolicy uint64         `json:"policy"`            // The policy for proposer selection
-	Deployer       common.Address `json:"contract-deployer"` // Address of the validator who deploys contract stored in bytecode
-	Bytecode       string         `json:"bytecode"`          // Bytecode of validators contract // would like this type to be []byte but the unmarshalling is not working
-	ABI            string         `json:"abi"`               // Bytecode of validators contract // ABI of the said contract
+	Epoch                  uint64                   `json:"epoch"`  // Epoch length to reset votes and checkpoint
+	ProposerPolicy         uint64                   `json:"policy"` // The policy for proposer selection
+	AutonityContractConfig *AutonityContractGenesis `json:"autonityContract"`
```
Comment (yazzaoui) [inline params/config.go]: I think it's better to have `AutonityContractGenesis` in `ChainConfig` directly

File: params/autonity_contract.go [hunk 2]
```diff
@@ -27,19 +28,45 @@ func (ut UserType) IsValid() bool {
...
 	// Address of the validator who deploys contract stored in bytecode
 	Deployer common.Address `json:"deployer" toml:",omitempty"`
-	// Bytecode of validators contract // would like this type to be []byte but the unmarshalling is not working
+	// Bytecode of validators contract
+	// would like this type to be []byte but the unmarshalling is not working
 	Bytecode string `json:"bytecode" toml:",omitempty"`
 	// Json ABI of the contract
 	ABI         string         `json:"abi "toml:",omitempty"`
-	MinGasPrice uint64         `json:"minGasPrice" toml:",omitempty"`
+	MinGasPrice int64          `json:"minGasPrice" toml:",omitempty"`
```
Comment (yazzaoui) [inline params/autonity_contract.go]: We need to be sure that `MinGasPrice` is always positive

**Comment by cmjc** [comment]
Looking at `autonity.sol` and:
>/*
>* RemoveValidator
>* Remove validator from validators list. function MUST be restricted to the Authority Account.
>*/
>function RemoveValidator(address _validator) public onlyValidators(msg.sender) {
>        require(validators.length > 1);

Why it is `onlyValidators(msg.sender)` and not `onlyOperator(msg.sender)`?

**Comment by yazzaoui** [comment]
> Looking at `autonity.sol` and:
> 
> > /*
> > 
> > * RemoveValidator
> > * Remove validator from validators list. function MUST be restricted to the Authority Account.
> >   */
> >   function RemoveValidator(address _validator) public onlyValidators(msg.sender) {
> >   require(validators.length > 1);
> 
> Why it is `onlyValidators(msg.sender)` and not `onlyOperator(msg.sender)`?

This PR is only for making the autonity contract interacting with the native client, the latest solidity file can be found here https://github.com/clearmatics/autonity/blob/autonity_contract_updates/contracts/autonity/contract/contracts/Autonity.sol
It's still work in progress and the function's signature might change so I prefer to wait a bit until we have something stable before writing the SRS.

File: params/config.go (hunk 1 above)
Comment (b00ris) [reply params/config.go]: The idea was to keep it separate from istanbul/tendermint config.

File: params/autonity_contract.go (hunk 2 above)
Comment (b00ris) [reply params/autonity_contract.go]: fixed

**Comment by yazzaoui** [comment]
LGTM

File: consensus/istanbul/backend/engine_test.go [hunk 3]
```diff
@@ -82,15 +84,23 @@ func getGenesisAndKeys(n int) (*core.Genesis, []*ecdsa.PrivateKey) {
 
 	// generate genesis block
 	genesis := core.DefaultGenesisBlock()
+	//genesis.Alloc=make(core.GenesisAlloc)
```
Comment (JekaMas) [inline consensus/istanbul/backend/engine_test.go]: Should it be removed?

File: consensus/istanbul/backend/engine_test.go [hunk 4]
```diff
@@ -82,15 +84,23 @@ func getGenesisAndKeys(n int) (*core.Genesis, []*ecdsa.PrivateKey) {
...
+	}
 	genesis.Config.Ethash = nil
 	genesis.Difficulty = defaultDifficulty
 	genesis.Nonce = emptyNonce.Uint64()
 	genesis.Mixhash = types.IstanbulDigest
 
 	AppendValidators(genesis, addrs)
+	err := genesis.Config.Istanbul.AutonityContractConfig.AddDefault().Validate()
+	if err != nil {
+		fmt.Println("consensus/istanbul/backend/engine_test.go:102 ", err)
```
Comment (JekaMas) [inline consensus/istanbul/backend/engine_test.go]: Is it a kind of debug?

**Pull request merged**